### PR TITLE
[PENDING] TW-134: Forward a single text message to contacts which have chat before

### PR DESCRIPTION
**Pull request by nqhhdev**
- Issue: #134 

- Resolved: 

- Has been chatted:

https://github.com/linagora/twake-on-matrix/assets/99852347/35856a54-8faf-4fd1-a2d3-8c3a56aa810a

- Never sent a message:

https://github.com/linagora/twake-on-matrix/assets/99852347/542292ca-30ee-4ef1-af78-272f82305c53

### CODE REVIEW

File: lib/pages/forward/forward.dart [hunk 1]
```diff
@@ -1,50 +1,81 @@
...
-
-  List<Room>? rooms;
+class ForwardController extends State<Forward> with ComparablePresentationContactMixin {
 
-  Timeline? timeline;
-
-  String? get roomId => context.vRouter.pathParameters['roomid'];
+  final _fetchContactsInteractor = getIt.get<FetchContactsInteractor>();
+  final streamController = StreamController<Either<Failure, GetContactsSuccess>>();
+  final networkStreamController = StreamController<Either<Failure, GetContactsSuccess>>();
```
Comment (hoangdat) [inline lib/pages/forward/forward.dart:30]: what is the different? why we need 2 different stream?

File: lib/pages/forward/forward.dart [hunk 2]
```diff
@@ -1,50 +1,81 @@
...
       );
     } else {
       setState(
-        () => selectedEvents.add(id),
+        () => selectedEvents.add(presentationForward),
       );
     }
     selectedEvents.sort(
-          (a, b) => a.compareTo(b),
+      (a, b) => a.id.compareTo(b.id),
```
Comment (hoangdat) [inline lib/pages/forward/forward.dart:78]: please more clear about the arguments

File: lib/pages/forward/presentation_forward.dart [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import 'package:equatable/equatable.dart';
+
+enum ForwardTypeEnum {
+  recently,
+  contacts,
+}
+class PresentationForward extends Equatable {
```
Comment (hoangdat) [inline lib/pages/forward/presentation_forward.dart:7]:
```suggestion
class ForwardToSelection extends Equatable {
```

File: lib/di/forward/forward_di.dart [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+import 'package:fluffychat/data/network/contact/tom_contact_api.dart';
+import 'package:fluffychat/data/repository/contact/tom_contact_repository_impl.dart';
+import 'package:fluffychat/di/base_di.dart';
+import 'package:fluffychat/domain/repository/contact_repository.dart';
+import 'package:fluffychat/domain/usecase/fetch_contacts_interactor.dart';
+import 'package:fluffychat/domain/usecase/lookup_contacts_interactor.dart';
+import 'package:get_it/get_it.dart';
+import 'package:matrix/matrix.dart';
+
+class ForwardDI extends BaseDI {
```
Comment (hoangdat) [inline lib/di/forward/forward_di.dart:12]: `ForwardToDI`

File: lib/pages/forward/forward.dart (hunk 2 above)
Comment (hoangdat) [reply lib/pages/forward/forward.dart:78]: `(thiz, that)`

Comment (hoangdat) [reply lib/pages/forward/forward.dart:78]: or `(current, next)`

File: lib/pages/forward/forward_view.dart [hunk 5]
```diff
@@ -31,32 +41,66 @@ class _ForwardViewState extends State<ForwardView> {
...
+    final roomId = await Matrix.of(context).client.startDirectChat(
+      widget.controller.selectedEvents.first.id.toTomMatrixId()
+    );
+
+    final hasRoom = rooms.any((element) => element.id == roomId);
+    if (hasRoom) {
+      final room = rooms.firstWhere((element) => element.id == roomId);
+      _sendForwardMessageAction(room);
+    } else {
+      await Matrix.of(context).client.getRoomById(roomId)!.leave();
```
Comment (hoangdat) [inline lib/pages/forward/forward_view.dart:65]: what is it? why `leave`?

File: lib/pages/forward/forward.dart (hunk 1 above)
Comment (nqhhdev) [reply lib/pages/forward/forward.dart:30]: Sorry, this is duplicated,  I'll remove itt

File: lib/pages/forward/forward_view.dart (hunk 5 above)
Comment (nqhhdev) [reply lib/pages/forward/forward_view.dart:65]: My point: 

1. When I selected one item and click forward then I got `roomId`.
2. If I call `Matrix.of(context).client.startDirectChat` => `Returns an existing direct room ID with this user or creates a new one.`

```
 final roomId = await Matrix.of(context).client.startDirectChat(
      widget.controller.selectedEvents.first.id.toTomMatrixId()
    );
```

3. If `roomId` existed in recent chat  => forward msg else => remove `roomId`.

I think we need to discuss with team about this

File: lib/pages/forward/forward.dart [hunk 6]
```diff
@@ -1,50 +1,81 @@
...
+
+  @override
+  void dispose() {
+    super.dispose();
+    streamController.close();
+    networkStreamController.close();
+  }
+
+  void fetchCurrentTomContacts() {
+    _fetchContactsInteractor
```
Comment (sherlockvn) [inline lib/pages/forward/forward.dart:53]: Why it's Contacts here ? 😕

File: lib/pages/forward/forward_view.dart [hunk 7]
```diff
@@ -31,32 +41,66 @@ class _ForwardViewState extends State<ForwardView> {
     });
   }
 
-  void forwardAction(BuildContext context) async {
-    final rooms = widget.controller.filteredRoomsForAll;
-    final room = rooms.firstWhere((element) => element.id == widget.controller.selectedEvents.first);
+  void _toggleContactChats() {
+    setState(() {
```
Comment (sherlockvn) [inline lib/pages/forward/forward_view.dart:45]: please don't use setState, use ValueNotifier and ValueListenableBuilder instead

File: lib/pages/forward/forward.dart (hunk 6 above)
Comment (nqhhdev) [reply lib/pages/forward/forward.dart:53]: It's get list contact broo

**Comment by hoangdat** [comment]
consider the UX

**Review by imGok — APPROVED** [review]

**Pull request closed without merging**